src/pTSL.py

```diff
@@ -1,36 +1,263 @@
+                            method='L-BFGS-B',
+                            args=(param, corpus_probs))
+
+        optimal_projection = dict(zip(param, proj_res.x))
+        self.probs.update(optimal_projection)
+
+    def print_projections(self):
+        for key in sorted(self.probs.keys()):
+            print("{}: {}".format(key, round(self.probs[key], 4)))
+
+    ''' 
+    initialize pTSL based on the given path to files (tier, prob, grammar)
+    '''
+    @classmethod
+    def from_file(cls, tier_file, factors_file, verbose=False):
+        '''
+            An alternative constructor that creates a pTSL object based on
+            the contents of input files. The files should have the following formats:
+
+            TIER_FILE
+            <symbol 1> <projection 1>
+            <symbol 2> <projection 2>
+            ...
+            <symbol n> <projection n>
+
+            FACTORS_FILE
+            <factor 1>
+            <factor 2>
+            ...
+            <factor n>
+
+
+            <factor i> is a segment separated by spaces.
+            Segments do not need to be single characters.
+
+            Input:
+                tier_file: A string specifying where to find the tier file.
+                factors_file: a string specifying where to find the factor file
+        '''
+
+        probs = {}
+        with open(tier_file, 'r') as csv_file:
+            csv_reader = csv.reader(csv_file)
+            for row in csv_reader:
+                if len(row) == 1:
+                    probs[row[0]] = 0
+                else:
+                    probs[row[0]] = float(row[1])
+
+        factors = []
+        with open(factors_file, 'r') as csv_file:
+            csv_reader = csv.reader(csv_file)
+            for row in csv_reader:
+                factors.append(tuple(row[0].split()))
+
+        return pTSL(probs, factors)
+
+    @staticmethod
+    def read_corpus_file(corpus_file, train=False):
+        '''
+            Read corpus data from a file
+            The file should have the following formats:
+
+            <string 1> <probability 1>
+            <string 2> <probability 1>
+            ...
+            <string n> <probability 1>
+
+
+            <string i> is a segment separated by spaces.
+            Segments do not need to be single characters.
+
+            probabilities are optional. They are used for the purpose of learning projection probabilities.
+
+            Input:
+                corpus_file: A string specifying where to find the corpus file
+                train:
+        '''
+        corpus_data = []
+        with open(corpus_file, 'r') as csv_file:
+            csv_reader = csv.reader(csv_file)
+            if train:
+                for row in csv_reader:
+                    corpus_data.append([row[0].split(), float(row[1])])
+            else:
+                for row in csv_reader:
+                    corpus_data.append(row[0].split())
+
+        return corpus_data

```

Comment:
It might be helpful to add a printed warning here if the length of every corpus string is 1 (which means the user probably forgot to put spaces in between segments).